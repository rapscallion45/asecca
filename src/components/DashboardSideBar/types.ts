/*
 * Dashboard Sidebar type definitions
 *
 * These type definitions are used in the applications Dashboard Sidebar component.
 */

/**
 * Nav Bar sidebar on close callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback IDashboardSidebarOnCloseCallback
 * @returns {void} - no return value
 */
export interface IDashboardSidebarOnCloseCallback {
  (): void;
}
