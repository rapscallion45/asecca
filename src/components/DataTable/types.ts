/*
 * Data Table type definitions
 *
 * These type definitions are used in the applications data table component.
 * Only make changes to this file if the corresponding change is made in the
 * Data Table component(s)
 */

/**
 * Data Table Row Cell Value types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof DataTable
 *
 * @typedef DataTableRowCellValue
 */
export type DataTableRowCellValue =
  | string
  | Array<string>
  | boolean
  | null
  | undefined;

/**
 * Data Table Column types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 * @memberof DataTable
 *
 * @typedef DataTableColumnType
 */
export type DataTableColumnType = 'string' | 'currency' | 'select' | 'check';

/**
 * Data Table Column properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableColumn
 * @prop {string} label - data table column name
 * @prop {string} key - data table column data identifier
 * @prop {DataTableColumnType} type - data table column type
 */
export interface IDataTableColumn {
  label: string;
  key: string;
  type: DataTableColumnType;
}

/**
 * Data Table Row properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableRow
 * @prop {string} label - data table row name
 */
export interface IDataTableRow {
  label: string;
}

/**
 * Data Table Cell Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCellValueCallback
 * @param {DataTableRowCellValue} value - updated cell value, can be null
 * @param {string} colKey - cell column data identifier
 * @param {number} rowIdx - cell row index in data table
 * @returns {void} - no return value
 */
export interface IDataTableEditCellValueCallback {
  (value: DataTableRowCellValue, colKey: string, rowIdx: number): void;
}

/**
 * Data Table Cell Value Retrieval callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableGetCellValueCallback
 * @param {number} rowIdx - cell row index in data table
 * @param {IDataTableColumn} column - cell column in data table
 * @returns {DataTableRowCellValue} - cell value, can be null or undefined
 */
export interface IDataTableGetCellValueCallback {
  (rowIdx: number, column: IDataTableColumn): DataTableRowCellValue;
}

/**
 * Data Table Currency Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCurrencyCellValueCallback
 * @param {string | null} value - updated cell value
 * @returns {void} - no return value
 */
export interface IDataTableEditCurrencyCellValueCallback {
  (value: string | null): void;
}

/**
 * Data Table Checkbox Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @callback IDataTableEditCheckboxCellValueCallback
 * @param {boolean | null} value - updated cell boolean value
 * @returns {void} - no return value
 */
export interface IDataTableEditCheckboxCellValueCallback {
  (value: boolean | null): void;
}
