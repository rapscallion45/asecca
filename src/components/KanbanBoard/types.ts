/*
 * Kanban Board type definitions
 *
 * These type definitions are used in the applications kanban board component.
 * Only make changes to this file if the corresponding change is made in the
 * Kanban Board component(s)
 */

/**
 * Kanban Board Column On Drop callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.7
 *
 * @callback IKanbanBoardColumnOnDropCallback
 * @param {number} colIndex - column index of task
 * @param {number} prevColIndex - previous column index of task
 * @param {number} taskIndex - index of task
 * @returns {void} - no return value
 */
export interface IKanbanBoardColumnOnDropCallback {
  (colIndex: number, prevColIndex: number, taskIndex: number): void;
}
