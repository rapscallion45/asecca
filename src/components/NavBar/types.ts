/*
 * Nav Bar type definitions
 *
 * These type definitions are used in the applications Nav Bar component.
 */

/**
 * Nav Bar sidebar open callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback INavBarOnSidebarOpenCallback
 * @returns {void} - no return value
 */
export interface INavBarOnSidebarOpenCallback {
  (): void;
}
