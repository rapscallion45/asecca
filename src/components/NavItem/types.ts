/*
 * Nav Item type definitions
 *
 * These type definitions are used in the applications Nav Item component.
 */

/**
 * Nav Item configuration properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof NavItem
 *
 * @typedef INavItemConfig
 * @prop {string} title - nav item title text
 * @prop {string} path - nav item URL path
 * @prop {any} icon - nav item icon
 * @prop {boolean} newTab - nav item link opens in new tab
 * @prop {Array<INavItemConfig>} children - nav item child components
 */
export interface INavItemConfig {
  title: string;
  path?: string;
  icon?: any;
  newTab?: boolean;
  children?: Array<INavItemConfig>;
}
