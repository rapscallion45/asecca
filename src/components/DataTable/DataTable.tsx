import { FC, Fragment } from 'react';
import { styled } from '@mui/material/styles';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell, { tableCellClasses } from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import Box from '@mui/material/Box';
import CircularProgress from '@mui/material/CircularProgress';
import Typography from '@mui/material/Typography';
import ErrorIcon from '@mui/icons-material/Error';
import DataTableRow from './DataTableRow/DataTableRow';
import ClientOnly from '../ClientOnly/ClientOnly';

/* table head cell stylings */
const StyledTableHeadCell = styled(TableCell)(({ theme }) => ({
  /* table head colors */
  [`&.${tableCellClasses.head}`]: {
    backgroundColor: theme.palette.secondary.main,
    color: theme.palette.primary.contrastText,
    fontSize: 16,
    fontWeight: 400,
  },
}));

interface DataTableProps {
  name: string;
  editColName: string;
  columns: any;
  rows: any;
  isLoading?: boolean;
  isError?: boolean;
}

/* Data Table */
/* ========== */
const DataTable: FC<DataTableProps> = (props) => {
  const {
    name,
    editColName,
    columns,
    rows,
    isLoading = false,
    isError = false,
  } = props;

  /* ClientOnly used to not allow tables to SSR */
  return (
    <ClientOnly>
      <TableContainer component={Paper}>
        <Table aria-label={`${name} table`}>
          <TableHead>
            <TableRow>
              {/* map passed columns to table headers row */}
              {columns.map((column: any) => (
                <StyledTableHeadCell key={column.key} align="center">
                  {column.label}
                </StyledTableHeadCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            <>
              {!isLoading && !isError && (
                <>
                  {/* map passed rows */}
                  {rows.map((row: any) => (
                    <Fragment key={row.name}>
                      <DataTableRow
                        row={row}
                        columns={columns}
                        /* we need the edit col to get 'Prevailing' value */
                        editCol={columns.find(
                          (col: any) => editColName === col.label
                        )}
                      />
                    </Fragment>
                  ))}
                </>
              )}
              {!isLoading && isError && (
                <Box
                  display="flex"
                  flexDirection="column"
                  py={7}
                  alignItems="center"
                >
                  <ErrorIcon color="error" fontSize="large" />
                  <Typography variant="h6" mt={2}>
                    Error loading the requested data
                  </Typography>
                </Box>
              )}
              {isLoading && (
                <Box
                  display="flex"
                  flexDirection="column"
                  py={7}
                  alignItems="center"
                >
                  <CircularProgress />
                  <Typography variant="h6" mt={2}>
                    Loading...
                  </Typography>
                </Box>
              )}
            </>
          </TableBody>
        </Table>
      </TableContainer>
    </ClientOnly>
  );
};

export default DataTable;
