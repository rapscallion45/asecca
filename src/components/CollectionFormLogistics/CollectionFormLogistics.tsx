import { FC, useState, useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import lodash from 'lodash';
import { AppState, AppDispatch } from '@/redux/store';
import {
  Card,
  CardHeader,
  CardContent,
  Box,
  Button,
  IconButton,
} from '@mui/material';
import { LoadingButton } from '@mui/lab';
import DeleteIcon from '@mui/icons-material/Delete';
import { ICollectionFormLogisticsData } from '@/lib/api/api-types';
import {
  getCollectionFormLogisticsCellValue,
  getCollectionFormLogisticsEditCellValue,
  getCollectionFormLogisticsTableColsList,
} from '@/utils';
import {
  saveByCollectionId as saveLogisticsByCollectionId,
  resetLogistics,
  editLogistics,
  addLogistics,
  deleteLogistics,
} from '@/redux/slices/collectionFormLogisticsSlice';
import DataTable from '@/components/DataTable/DataTable';
import {
  DataTableRowCellValue,
  IDataTableColumn,
} from '@/components/DataTable/types';
import columns from './collectionFormLogisticsTableColumns';

/**
 * Collection Form Logistics Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @typedef ICollectionFormLogisticsProps
 * @prop {string} collectionId - ID string of Collection for table data API call
 */
interface ICollectionFormLogisticsProps {
  collectionId: string;
}

/**
 * Collection Form Logistics
 *
 * Presents the Collection Form Logisitcs table to the user, populated with data
 * fetched from API: /api/collection/logistics/api/logistics
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @component
 * @param {ICollectionFormLogisticProps} props - component props
 * @returns {FC} - collection form logistics table functional component
 */
const CollectionFormLogistics: FC<ICollectionFormLogisticsProps> = (props) => {
  const { collectionId } = props;

  /* shorthand helper for dispatching redux actions */
  const dispatch = useDispatch<AppDispatch>();

  /* get collection form logisitcs data held in redux state */
  const {
    data,
    loading,
    error,
    saving,
    edited,
    types: logisticsTypes,
  } = useSelector((state: AppState) => state.collectionFormLogistics);

  /**
   * Table column list
   *
   * Local state of columns list for available logistics and facilities
   *
   * @since 0.0.13
   *
   * @constant
   */
  const [colList, setColList] = useState<Array<IDataTableColumn>>([]);

  /* Whenever Logisitcs data updates, update column list */
  useEffect(() => {
    setColList(
      getCollectionFormLogisticsTableColsList(columns, data, logisticsTypes)
    );
  }, [data, logisticsTypes]);

  /**
   * Handles the saving of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   */
  const handleSave = useCallback(() => {
    dispatch(
      saveLogisticsByCollectionId({
        data: {
          collection: collectionId,
          rows: data?.rows,
        },
      })
    );
  }, [collectionId, data?.rows, dispatch]);

  /**
   * Handles the resetting of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   */
  const handleReset = useCallback(() => {
    dispatch(resetLogistics());
  }, [dispatch]);

  /**
   * Handles the update of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   * @param {DataTableRowCellValue} value - value to be used in update
   * @param {string} colKey - column key to be updated
   * @param {number} rowIdx - table row index to be updated
   */
  const handleEditCellValue = useCallback(
    (value: DataTableRowCellValue, colKey: string, rowIdx: number) => {
      dispatch(
        editLogistics(
          getCollectionFormLogisticsEditCellValue(
            data,
            value,
            colKey,
            rowIdx,
            logisticsTypes.logistics_types,
            colList
          )
        )
      );
    },
    [data, colList, logisticsTypes, dispatch]
  );

  /**
   * Handles any required logic when determining a cell's display value
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   * @param {number} rowIdx - table row index to get value from
   * @param {IDataTableColumn} column - column to get value from
   * @return {DataTableRowCellValue} - cell value, can be null or undefined
   */
  const handleGetCellValue = useCallback(
    (rowIdx: number, column: IDataTableColumn): DataTableRowCellValue =>
      getCollectionFormLogisticsCellValue(
        data?.rows[rowIdx],
        column,
        logisticsTypes.logistics_types
      ),
    [data?.rows, logisticsTypes.logistics_types]
  );

  /**
   * Determines whether specific table cell can be edited
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   * @param {string} colKey - column key to be updated
   * @param {number} rowIdx - table row index to be updated
   * @returns {boolean} - whether cell is editable
   */
  const canEditCellValue = useCallback(
    (colKey: string, rowIdx: number): boolean => {
      /* logistics type can always be changed */
      if (colKey === 'logistics_type') return true;
      return (
        logisticsTypes.logistics_types
          .find(
            (type) => type.logistics_type === data.rows[rowIdx].logistics_type
          )
          ?.compatible_facilities.some(
            (facility) => lodash.snakeCase(facility) === colKey
          ) || false
      );
    },
    [data, logisticsTypes.logistics_types]
  );

  /**
   * Handles addition of table row for new Logistics type
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   */
  const handleAddRow = useCallback(() => dispatch(addLogistics()), [dispatch]);

  /**
   * Handles deletion of table row for passed row index
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   * @param {number} rowIdx - table row index to be deleted
   */
  const handleDeleteRow = useCallback(
    (rowIdx: number) => dispatch(deleteLogistics({ rowIdx })),
    [dispatch]
  );

  /**
   * Handles requests for the action component for specific column
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   * @param {string} colKey - table column key of the action cell
   * @param {number} rowIdx - table row index of the action cell
   * @returns {ReactNode} - action component to present
   */
  const getActionComponent = (colKey: string, rowIdx: number) => (
    <IconButton aria-label="delete" onClick={() => handleDeleteRow(rowIdx)}>
      <DeleteIcon />
    </IconButton>
  );

  return (
    <Card>
      <CardHeader title="Logistics" />
      <CardContent sx={{ pt: 0 }}>
        <DataTable
          name="collection form logistics"
          /* filter table columns by current logistics type */
          columns={colList}
          /* can edit all cells in logistics table */
          editableColLabels={colList.map((col) => col.label)}
          /* build table row props from logistics data */
          rows={data?.rows.map((logistic: ICollectionFormLogisticsData) => ({
            label: logistic.logistics_type,
          }))}
          isLoading={loading}
          error={error}
          editCellValueCallback={handleEditCellValue}
          getCellValueCallback={handleGetCellValue}
          canEditCellValueCallback={canEditCellValue}
          getActionComponent={getActionComponent}
        />
        <Box
          sx={{
            marginTop: 2,
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'left',
          }}
        >
          <Button
            color="secondary"
            variant="contained"
            onClick={handleAddRow}
            disabled={saving || loading}
          >
            Add +
          </Button>
        </Box>
        <Box
          sx={{
            marginTop: 2,
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'left',
          }}
        >
          <Button
            color="secondary"
            variant="outlined"
            onClick={handleReset}
            disabled={saving || loading || !edited}
            sx={{ backgroundColor: 'common.white' }}
          >
            Reset
          </Button>
          <LoadingButton
            color="secondary"
            variant="contained"
            onClick={handleSave}
            disabled={saving || loading || !edited}
            loading={saving}
            sx={{ ml: 2 }}
          >
            Save
          </LoadingButton>
        </Box>
      </CardContent>
    </Card>
  );
};

export default CollectionFormLogistics;
