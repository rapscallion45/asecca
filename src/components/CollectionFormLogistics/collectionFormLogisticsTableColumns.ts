import { IDataTableColumn } from '@/components/DataTable/types';

/**
 * Collection Form Logistics table column defintions
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof CollectionFormLogistics
 *
 * @constant
 * @type {Array<IDataTableColumn>}
 */
const collectionFormLogisticsTableColumns: Array<IDataTableColumn> = [
  { label: 'Logistics Type', key: 'logistics_type', type: 'string' },
];

export default collectionFormLogisticsTableColumns;
