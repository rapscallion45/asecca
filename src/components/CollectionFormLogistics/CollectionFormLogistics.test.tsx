import React from 'react';
import { Provider } from 'react-redux';
import {
  fireEvent,
  render,
  screen,
  waitFor,
  act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import { fetchByCollectionId as fetchCollectionFormLogisticsByCollectionId } from '@/redux/slices/collectionFormLogisticsSlice';
import { ICollectionFormLogisticsTypesData } from '@/lib/api/api-types';
import store from '../../redux/store';
import columns from './collectionFormLogisticsTableColumns';
import CollectionFormLogistics from './CollectionFormLogistics';
import collectionFormLogisticsDataMock from '../../../__mocks__/CollectionForm/collectionFormLogisticsDataMock';
import collectionFormLogisticsTypesDataMock from '../../../__mocks__/CollectionForm/collectionFormLogisticsTypesDataMock';
import { IDataTableColumn } from '../DataTable/types';

/* default test query ID */
const query: string = '66135000001760012';

/* chosen data row index to check during tests */
const testDataRowIdx = 1;

/* test Logisitics Type index to check during tests */
const testTypeIdx = 0;

/**
 * Collection Form Logistics Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Collection Form Logistics', () => {
  it('Renders correctly', async () => {
    /** Arrange */
    store.dispatch(
      fetchCollectionFormLogisticsByCollectionId({
        collectionId: query,
      })
    );

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <CollectionFormLogistics collectionId={query} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Table Callbacks', () => {
    beforeAll(() => jest.spyOn(window, 'fetch'));

    it('Should call API with updated values when value updated and save button clicked', async () => {
      /** Arrange */
      const testSelectDropdownIdx = 0;
      act(() => {
        store.dispatch(
          fetchCollectionFormLogisticsByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Assert - get API to have been called */
      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(1);
      });
      expect(window.fetch).toHaveBeenCalledWith(
        `/api/collection/logistics/api/logistics?collection=${query}`,
        expect.objectContaining({
          method: 'GET',
          headers: { 'Content-Type': 'application/json' },
        })
      );
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormLogistics collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        store
          .getState()
          .collectionFormLogistics.types.logistics_types.forEach(
            (type: ICollectionFormLogisticsTypesData) => {
              type.compatible_facilities.forEach((compatibleFacility: string) =>
                expect(screen.getByText(compatibleFacility)).toBeInTheDocument()
              );
            }
          );
      });
      /* check data has been rendered */
      expect(
        screen.getByDisplayValue(
          collectionFormLogisticsDataMock.rows[testDataRowIdx].logistics_type
        )
      ).toBeInTheDocument();
      /** Act - click on dropdown, update value, click Save button */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormLogisticsDataMock.rows[testDataRowIdx].logistics_type
          ),
          {
            target: {
              value:
                collectionFormLogisticsTypesDataMock.logistics_types[
                  testSelectDropdownIdx
                ].logistics_type,
            },
          }
        );
      });
      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(
          screen.getByDisplayValue(
            collectionFormLogisticsTypesDataMock.logistics_types[
              testSelectDropdownIdx
            ].logistics_type
          )
        ).toBeInTheDocument();
      });
      /** Act - click Save button */
      act(() => {
        fireEvent(
          screen.getByText('Save'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - save API to have been called with data from state */
      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(2);
      });
      expect(window.fetch).toHaveBeenCalledWith(
        '/api/collection/logistics/api/logistics',
        expect.objectContaining({
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            collection: query,
            rows: store.getState().collectionFormLogistics.data.rows,
          }),
        })
      );
    });

    it('Should reset table data when cancel button clicked', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormLogisticsByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormLogistics collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        store
          .getState()
          .collectionFormLogistics.types.logistics_types.forEach(
            (type: ICollectionFormLogisticsTypesData) => {
              type.compatible_facilities.forEach((compatibleFacility: string) =>
                expect(screen.getByText(compatibleFacility)).toBeInTheDocument()
              );
            }
          );
      });
      expect(
        screen.getByDisplayValue(
          collectionFormLogisticsDataMock.rows[testDataRowIdx].logistics_type
        )
      ).toBeInTheDocument();
      /** Act - click on dropdown, update value */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormLogisticsDataMock.rows[testDataRowIdx].logistics_type
          ),
          {
            target: {
              value:
                collectionFormLogisticsTypesDataMock.logistics_types[
                  testTypeIdx
                ].logistics_type,
            },
          }
        );
      });
      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(
          screen.getByDisplayValue(
            collectionFormLogisticsTypesDataMock.logistics_types[testTypeIdx]
              .logistics_type
          )
        ).toBeInTheDocument();
      });
      /** Act - click Reset button */
      act(() => {
        fireEvent(
          screen.getByText('Reset'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - table data should no longer hold updated values and be reset */
      await waitFor(() => {
        expect(
          screen.queryByDisplayValue(
            collectionFormLogisticsTypesDataMock.logistics_types[testTypeIdx]
              .logistics_type
          )
        ).toBeNull();
      });
      expect(
        screen.getByDisplayValue(
          collectionFormLogisticsDataMock.rows[testDataRowIdx].logistics_type
        )
      ).toBeInTheDocument();
    });

    it('Should add row to table when "Add +" button clicked', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormLogisticsByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormLogistics collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        store
          .getState()
          .collectionFormLogistics.types.logistics_types.forEach(
            (type: ICollectionFormLogisticsTypesData) => {
              type.compatible_facilities.forEach((compatibleFacility: string) =>
                expect(screen.getByText(compatibleFacility)).toBeInTheDocument()
              );
            }
          );
      });
      expect(
        screen.getByDisplayValue(
          collectionFormLogisticsDataMock.rows[testDataRowIdx].logistics_type
        )
      ).toBeInTheDocument();
      /** Act - click on "Add +" button */
      act(() => {
        fireEvent(
          screen.getByText('Add +'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - wait for data to update */
      await waitFor(() => {
        expect(
          screen.getByDisplayValue(
            collectionFormLogisticsTypesDataMock.logistics_types[testTypeIdx]
              .logistics_type
          )
        ).toBeInTheDocument();
      });
    });

    it('Should delete row from table when delete icon button clicked', async () => {
      /** Arrange */
      const deleteRowIdx = 0;
      act(() => {
        store.dispatch(
          fetchCollectionFormLogisticsByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormLogistics collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        store
          .getState()
          .collectionFormLogistics.types.logistics_types.forEach(
            (type: ICollectionFormLogisticsTypesData) => {
              type.compatible_facilities.forEach((compatibleFacility: string) =>
                expect(screen.getByText(compatibleFacility)).toBeInTheDocument()
              );
            }
          );
      });
      expect(
        screen.getByDisplayValue(
          collectionFormLogisticsDataMock.rows[testDataRowIdx].logistics_type
        )
      ).toBeInTheDocument();
      /** Act - click on delete icon button of first row */
      act(() => {
        fireEvent(
          screen.getAllByTestId('DeleteIcon')[deleteRowIdx],
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - wait for data to update */
      await waitFor(() => {
        expect(
          screen.queryByDisplayValue(
            collectionFormLogisticsDataMock.rows[testDataRowIdx].logistics_type
          )
        ).toBeNull();
      });
    });
  });

  describe('Table Columns', () => {
    it('Should render correctly for passed permission', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormLogisticsByCollectionId({
            collectionId: query,
          })
        );
      });

      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormLogistics collectionId={query} />
        </Provider>
      );

      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'delete_row')
            expect(screen.queryByText(column.label)).toBeInTheDocument();
        });
      });
    });
  });
});
