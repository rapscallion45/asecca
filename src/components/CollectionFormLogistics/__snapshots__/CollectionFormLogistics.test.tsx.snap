// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Form Logistics Renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
      >
        Logistics
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root css-1hizqsf-MuiCardContent-root"
  >
    <div
      className="MuiBox-root css-kh5ii8"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Add +
      </button>
    </div>
    <div
      className="MuiBox-root css-kh5ii8"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1ex1hsz-MuiButtonBase-root-MuiButton-root"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Reset
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1fddvdp-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        disabled={true}
        id=":r0:"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Save
      </button>
    </div>
  </div>
</div>
`;
